homework_solution.md

```diff
@@ -0,0 +1,474 @@
+    ------------------------------------------+-------------------
+    Bikeshare Ridership (2017 Q1).csv        | 10/1/2017 0:03
+    Bikeshare Ridership (2017 Q2).csv        | 10/4/2017 0:00
+    Bikeshare Ridership (2017 Q3).csv        | 7/10/2017 0:00
+    Bikeshare Ridership (2017 Q4).csv        | 10/01/17 00:00:01
+    Bike Share Toronto Ridership_Q1 2018.csv | 1/10/2018 0:01
+    Bike Share Toronto Ridership_Q2 2018.csv | 4/10/2018 0:01
+    Bike Share Toronto Ridership_Q3 2018.csv | 7/10/2018 0:00
+    Bike Share Toronto Ridership_Q4 2018.csv | 10/10/2018 0:01
+    ```
+
+    We can see there that:
+
+    - Throughout 2017 there's no consistency in using `dd/mm` and `mm/dd`. The first 2 quarters use `dd/m`, while the last 2 use `m/dd`.
+    - Further, throughout 2017 there's no consistency in using `yy` and `yyyy`. The first 3 quarters use the full year, while the last one uses short year.
+    - Time format is also inconsistent throughout 2017, with the first 3 quarters using `h:mm`, while the last one uses `hh:mm:ss`
+    - 2018 was a beautiful year for date formating in the bike share organization
+
+2.  _**Take a look at Postgres's [date functions](https://www.postgresql.org/docs/12/functions-datetime.html), and fill in the missing date data using proper timestamps. You may have to write several queries to do this.**_
+
+    Our final goal is to get all dates to be in the format used throughout 2018 files, which is `mm/dd/yyyy h:mm`.
+
+    Because each quarter may have a different format, and because there's no consistency in the original_filenames, we'll run separate queries for each quarter, using `to_timestamp` with the format we saw for each file.
+
+    ## Steps
+
+    1. Check if there are cases where start time is present without end time or vice-versa. That'll determine whether we can combine updates for both columns or need to run them separately.
+
+       ```
+       SELECT COUNT(*)
+       FROM trips
+       WHERE
+          (start_time_str IS NULL AND end_time_str IS NOT NULL)
+       OR
+          (start_time_str IS NOT NULL AND end_time_str IS NULL);
+       ```
+
+       The result is `0`, so we can rely on checking for nulls on either column and trust that it accounts for both.
+
+    2. Update `2017 Q1` and `2017 Q2` records, since they share the same format
+
+       ```
+       UPDATE trips
+       SET
+          start_time = (SELECT TO_TIMESTAMP(start_time_str, 'DD/MM/YYYY HH24:MI')),
+          end_time = (SELECT TO_TIMESTAMP(end_time_str, 'DD/MM/YYYY HH24:MI'))
+       WHERE original_filename SIMILAR TO '%(2017 Q1|2017 Q2)%'
+       AND start_time_str IS NOT NULL;
+       ```
+
+    3. Update `2017 Q3` records
+
+       ```
+       UPDATE trips
+       SET
+          start_time = (SELECT TO_TIMESTAMP(start_time_str, 'MM/DD/YYYY HH24:MI')),
+          end_time = (SELECT TO_TIMESTAMP(end_time_str, 'MM/DD/YYYY HH24:MI'))
+       WHERE original_filename LIKE '%2017 Q3%'
```

Comment:
[suggestion]
If it's just the one file, why not target it explicitly?

Reply:
You mean, with the full name?
I didn't think too much about performance (maybe I should have), and I'm not sure about the cost difference in running it with the wildcards vs. without.
For me here it was more about clarity, and straight up being able to see the portion of the file name that was relevant (ie. the year and quarter).

Reply:
Performance is an important consideration, but it's almost never the most important one in my experience (which, admittedly, does not extend to many industries).  You're absolutely right to prioritize readability.

This also has the added benefit that if the filenames were to change somehow, this would keep on working!